config/initializers/doorkeeper.rb

```diff
@@ -1,23 +1,46 @@
 # frozen_string_literal: true
 
+require "./lib/level_of_authentication"
+
 Doorkeeper.configure do
   # Change the ORM that doorkeeper will use (requires ORM extensions installed).
   # Check the list of supported ORMs here: https://github.com/doorkeeper-gem/doorkeeper#orms
   orm :active_record
 
   # This block will be called to check whether the resource owner is authenticated or not.
   resource_owner_authenticator do
+    if Rails.application.config.feature_flag_enforce_levels_of_authentication
+      level_scopes = request.params[:scope]
+                            .split(" ")
+                            .select { |scope| scope.starts_with?("level") }
+
+      required_auth_level = LevelOfAuthentication.select_highest_level(level_scopes)
+      current_auth_level = session[:level_of_authentication].to_s
+    else
+      required_auth_level = "level1"
+      current_auth_level = "level1"
+    end
+
     if current_user
+      unless LevelOfAuthentication.current_auth_greater_or_equal_to_required(current_auth_level, required_auth_level)
+        raise LevelOfAuthenticationTooLowError
+      end
+
       current_user
     else
+      session[:required_authtication_level] = required_auth_level
```

Comment:
This has a typo, but it's also not used anywhere.

Reply:
ah yes! I can drop this we went for params instead!